File: src/harness/capabilities.ts
import { Queue } from "../comms/queue";
import { SharedFileSystem } from "../comms/fs";

// #EXT-005-REQ-3 Start
/**
 * Capability-scoped I/O handles.
 *
 * An agent never receives a raw `Queue` or `SharedFileSystem`. Instead the
 * harness mints narrow, revocable *handles* — each handle exposes exactly one
 * verb (send / receive / write / read) bound to one underlying resource. The
 * agent holds the handle, not the resource, so it has no path back to the raw
 * module, sibling queues, or other file-system paths.
 *
 * Capabilities are *revocable*: on teardown the harness calls {@link revoke},
 * after which every granted handle fails closed — any use throws a
 * {@link RevokedCapabilityError} and performs no side effect.
 */

/** The four capability kinds an agent can be granted. */
export type CapabilityKind =
  | "QueueSend"
  | "QueueReceive"
  | "FsWrite"
  | "FsRead";

/** Raised when a handle is used after its grants have been revoked. */
export class RevokedCapabilityError extends Error {
  /** The capability kind that was attempted after revocation. */
  readonly capability: CapabilityKind;

  constructor(capability: CapabilityKind) {
    super(
      `Capability "${capability}" has been revoked; the handle can no longer be used.`
    );
    this.name = "RevokedCapabilityError";
    this.capability = capability;
    Object.setPrototypeOf(this, RevokedCapabilityError.prototype);
  }
}

/** A handle that lets an agent enqueue onto exactly one queue. */
export interface QueueSendHandle {
  readonly kind: "QueueSend";
  send(msg: unknown): void;
}

/** A handle that lets an agent dequeue from exactly one queue. */
export interface QueueReceiveHandle {
  readonly kind: "QueueReceive";
  receive(): Promise<unknown>;
}

/** A handle that lets an agent write under exactly one shared file system. */
export interface FsWriteHandle {
  readonly kind: "FsWrite";
  write(relativePath: string, data: string): void;
}

/** A handle that lets an agent read under exactly one shared file system. */
export interface FsReadHandle {
  readonly kind: "FsRead";
  read(relativePath: string): string;
}

/** Union of every handle type an agent may hold. */
export type CapabilityHandle =
  | QueueSendHandle
  | QueueReceiveHandle
  | FsWriteHandle
  | FsReadHandle;

/**
 * The minimal bundle of handles an agent is constructed with. Any absent field
 * means the agent has *no* access to that capability — there is no ambient
 * fallback to a global queue/fs/network.
 */
export interface Grants {
  readonly queueSend?: QueueSendHandle;
  readonly queueReceive?: QueueReceiveHandle;
  readonly fsWrite?: FsWriteHandle;
  readonly fsRead?: FsReadHandle;
}

/**
 * Declarative request for capabilities, naming the underlying resources to bind
 * each handle to. The harness owns the resources; the agent only names which of
 * the *offered* capabilities it should receive.
 */
export interface GrantSpec {
  /** Bind a send handle to this queue. */
  readonly queueSend?: Queue<unknown>;
  /** Bind a receive handle to this queue. */
  readonly queueReceive?: Queue<unknown>;
  /** Bind write/read handles to this shared file system. */
  readonly fs?: SharedFileSystem;
  /** Which fs verbs to grant when `fs` is supplied (default: none). */
  readonly fsWrite?: boolean;
  readonly fsRead?: boolean;
}

/**
 * A {@link Grants} bundle plus the means to revoke it. The harness keeps the
 * `revoke` function private to itself; agents only ever see {@link Grants}.
 */
export interface RevocableGrants {
  readonly grants: Grants;
  /** Invalidate every handle in the bundle. Idempotent. */
  revoke(): void;
}

/**
 * Mint a {@link RevocableGrants} bundle from a {@link GrantSpec}.
 *
 * Each requested handle closes over a shared `revoked` flag; once {@link revoke}
 * flips it, every handle throws {@link RevokedCapabilityError} on use before
 * touching the underlying resource. Handles wrap the resource so the agent can
 * never reach the raw `Queue`/`SharedFileSystem` instance.
 */
export function grant(spec: GrantSpec): RevocableGrants {
  const state = { revoked: false };

  const guard = (kind: CapabilityKind): void => {
    if (state.revoked) {
      throw new RevokedCapabilityError(kind);
    }
  };

  const grants: {
    queueSend?: QueueSendHandle;
    queueReceive?: QueueReceiveHandle;
    fsWrite?: FsWriteHandle;
    fsRead?: FsReadHandle;
  } = {};

  if (spec.queueSend) {
    const q = spec.queueSend;
    grants.queueSend = Object.freeze({
      kind: "QueueSend" as const,
      send(msg: unknown): void {
        guard("QueueSend");
        q.enqueue(msg);
      },
    });
  }

  if (spec.queueReceive) {
    const q = spec.queueReceive;
    grants.queueReceive = Object.freeze({
      kind: "QueueReceive" as const,
      async receive(): Promise<unknown> {
        guard("QueueReceive");
        return q.dequeue();
      },
    });
  }

  if (spec.fs && spec.fsWrite) {
    const fs = spec.fs;
    grants.fsWrite = Object.freeze({
      kind: "FsWrite" as const,
      write(relativePath: string, data: string): void {
        guard("FsWrite");
        fs.write(relativePath, data);
      },
    });
  }

  if (spec.fs && spec.fsRead) {
    const fs = spec.fs;
    grants.fsRead = Object.freeze({
      kind: "FsRead" as const,
      read(relativePath: string): string {
        guard("FsRead");
        return fs.read(relativePath);
      },
    });
  }

  const frozen = Object.freeze(grants);

  return {
    grants: frozen,
    revoke(): void {
      state.revoked = true;
    },
  };
}

/**
 * Convenience wrapper to revoke a {@link RevocableGrants} bundle. After this,
 * every handle in the bundle fails closed.
 */
export function revoke(grants: RevocableGrants): void {
  grants.revoke();
}
// #EXT-005-REQ-3 End

File: jaros/harness/capabilities.py
"""Capability/handle model for the Architectural Harness (EXT-005 / REQ-3, REQ-6).

Agents hold no ambient power. They receive only the specific, scoped handles the
harness grants them; those handles wrap the underlying :class:`~jaros.comms.queue.Queue`
and :class:`~jaros.comms.fs.SharedFileSystem` so an agent can never reach the raw
modules, other queues, other paths, or the network.

Security boundary (EXT-005 / REQ-6): capability scoping here is **structural
least-privilege** for correctness and blast-radius control — a bug or bad
decision cannot touch what it was never granted. It is *not* an adversarial
sandbox against hostile code sharing the interpreter; real isolation against
hostile code is the host's job (process, container, VPC).

A capability is a small, frozen descriptor (``QueueSend``, ``QueueReceive``,
``FsWrite``, ``FsRead``). A :class:`GrantSpec` bundles the capabilities a single
agent is allowed, together with the concrete backing objects. :func:`grant`
turns a spec into a frozen :class:`Grants` bundle of scoped handles; :func:`revoke`
invalidates them so any subsequent use raises :class:`RevokedCapabilityError`
*before* any side effect occurs.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from jaros.comms.fs import SharedFileSystem
    from jaros.comms.queue import Queue


# #EXT-005-REQ-3 Start
class CapabilityError(Exception):
    """Base class for capability/handle errors."""


class RevokedCapabilityError(CapabilityError):
    """Raised when a revoked handle is used, before any side effect occurs."""


# --- Capability kinds -------------------------------------------------------
#
# Each capability kind is a frozen marker describing one permitted action class.
# They are deliberately tiny and immutable: agents may hold references to them
# but can neither mutate them nor manufacture new powers from them.


@dataclass(frozen=True, slots=True)
class QueueSend:
    """Capability to enqueue onto a specific queue."""

    name: str = "queue.send"


@dataclass(frozen=True, slots=True)
class QueueReceive:
    """Capability to dequeue/peek from a specific queue."""

    name: str = "queue.receive"


@dataclass(frozen=True, slots=True)
class FsWrite:
    """Capability to write to the shared file system."""

    name: str = "fs.write"


@dataclass(frozen=True, slots=True)
class FsRead:
    """Capability to read from the shared file system."""

    name: str = "fs.read"


Capability = QueueSend | QueueReceive | FsWrite | FsRead


# --- Scoped handles ---------------------------------------------------------
#
# Handles are the *only* objects an agent ever touches. They wrap the backing
# queue/fs, expose a minimal surface, honour revocation, and are frozen so the
# agent cannot swap out the backing object or otherwise reach the raw module.


class _RevocableHandle:
    """Mixin providing frozen attributes + a shared revocation flag.

    A handle is frozen after construction (``__setattr__``/``__delattr__`` raise)
    so an agent cannot rebind ``_backing`` to a different queue/fs or clear the
    revocation flag. Revocation flips a one-element mutable list shared with the
    owning :class:`Grants` bundle, so revoking the bundle revokes every handle.
    """

    _frozen: bool = False

    def _check_live(self) -> None:
        # ``_revoked`` is a single-element list shared across a bundle.
        if self._revoked[0]:
            raise RevokedCapabilityError(
                f"capability handle has been revoked: {type(self).__name__}"
            )

    def __setattr__(self, name: str, value: Any) -> None:
        if getattr(self, "_frozen", False):
            raise CapabilityError(
                f"capability handles are frozen; cannot set {name!r}"
            )
        object.__setattr__(self, name, value)

    def __delattr__(self, name: str) -> None:
        raise CapabilityError("capability handles are frozen; cannot delete attributes")


class QueueSendHandle(_RevocableHandle):
    """Scoped send-only view of a backing :class:`~jaros.comms.queue.Queue`."""

    def __init__(self, backing: Queue[Any], revoked: list[bool]) -> None:
        object.__setattr__(self, "_backing", backing)
        object.__setattr__(self, "_revoked", revoked)
        object.__setattr__(self, "_frozen", True)

    def send(self, msg: Any) -> None:
        self._check_live()
        self._backing.enqueue(msg)


class QueueReceiveHandle(_RevocableHandle):
    """Scoped receive-only view of a backing :class:`~jaros.comms.queue.Queue`."""

    def __init__(self, backing: Queue[Any], revoked: list[bool]) -> None:
        object.__setattr__(self, "_backing", backing)
        object.__setattr__(self, "_revoked", revoked)
        object.__setattr__(self, "_frozen", True)

    def receive(self) -> Any:
        self._check_live()
        return self._backing.dequeue()

    def peek(self) -> Any:
        self._check_live()
        return self._backing.peek()


class FsWriteHandle(_RevocableHandle):
    """Scoped write-only view of a backing :class:`~jaros.comms.fs.SharedFileSystem`."""

    def __init__(self, backing: SharedFileSystem, revoked: list[bool]) -> None:
        object.__setattr__(self, "_backing", backing)
        object.__setattr__(self, "_revoked", revoked)
        object.__setattr__(self, "_frozen", True)

    def write(self, path: str, data: str) -> None:
        self._check_live()
        self._backing.write(path, data)


class FsReadHandle(_RevocableHandle):
    """Scoped read-only view of a backing :class:`~jaros.comms.fs.SharedFileSystem`."""

    def __init__(self, backing: SharedFileSystem, revoked: list[bool]) -> None:
        object.__setattr__(self, "_backing", backing)
        object.__setattr__(self, "_revoked", revoked)
        object.__setattr__(self, "_frozen", True)

    def read(self, path: str) -> str:
        self._check_live()
        return self._backing.read(path)


# --- Grant spec + bundle ----------------------------------------------------

#: Mapping of role names to their permitted capability types.
BUILTIN_ROLES: dict[str, tuple[type[Capability], ...]] = {
    "AdminRole": (FsRead, FsWrite, QueueSend, QueueReceive),
    "AnalystRole": (FsRead,),
    "ReporterRole": (FsWrite, QueueSend),
    "GuestRole": (FsRead,),
    "QueueSendRole": (QueueSend,),
    "QueueReceiveRole": (QueueReceive,),
    "FsWriteRole": (FsWrite,),
    "FsReadRole": (FsRead,),
    "FsRole": (FsRead, FsWrite),
}


@dataclass(frozen=True, slots=True)
class GrantSpec:
    """Declarative request for the capabilities a single agent should hold under a role.

    The harness (never the agent) supplies this. It names the role the agent
    is assigned to, and the concrete backing objects to scope them to.
    """

    role: str
    queue: Queue[Any] | None = None
    fs: SharedFileSystem | None = None

    def has(self, cap_type: type) -> bool:
        caps = BUILTIN_ROLES.get(self.role, ())
        return cap_type in caps


@dataclass(frozen=True, slots=True)
class Grants:
    """Immutable bundle of the scoped handles an agent actually holds.

    Only the handles populated here are reachable by the agent. ``_revoked`` is
    the shared revocation flag; :func:`revoke` flips it, invalidating every
    handle at once. The agent never receives a reference to the raw queue/fs.
    """

    queue_send: QueueSendHandle | None = None
    queue_receive: QueueReceiveHandle | None = None
    fs_write: FsWriteHandle | None = None
    fs_read: FsReadHandle | None = None
    role: str = ""
    _revoked: list[bool] = field(default_factory=lambda: [False])

    @property
    def revoked(self) -> bool:
        return self._revoked[0]


def grant(spec: GrantSpec) -> Grants:
    """Produce a frozen :class:`Grants` bundle of scoped handles from ``spec``.

    Only capabilities named in ``spec`` (and backed by a supplied queue/fs)
    yield a handle; everything else is ``None`` and thus unreachable.

    Raises:
        CapabilityError: if a capability is requested without its backing object.
    """
    revoked: list[bool] = [False]

    queue_send = queue_receive = fs_write = fs_read = None

    if spec.has(QueueSend):
        if spec.queue is None:
            raise CapabilityError("QueueSend requires a backing queue")
        queue_send = QueueSendHandle(spec.queue, revoked)
    if spec.has(QueueReceive):
        if spec.queue is None:
            raise CapabilityError("QueueReceive requires a backing queue")
        queue_receive = QueueReceiveHandle(spec.queue, revoked)
    if spec.has(FsWrite):
        if spec.fs is None:
            raise CapabilityError("FsWrite requires a backing file system")
        fs_write = FsWriteHandle(spec.fs, revoked)
    if spec.has(FsRead):
        if spec.fs is None:
            raise CapabilityError("FsRead requires a backing file system")
        fs_read = FsReadHandle(spec.fs, revoked)

    return Grants(
        queue_send=queue_send,
        queue_receive=queue_receive,
        fs_write=fs_write,
        fs_read=fs_read,
        role=spec.role,
        _revoked=revoked,
    )


def revoke(grants: Grants) -> None:
    """Invalidate every handle in ``grants``.

    After this returns, any use of any handle raises
    :class:`RevokedCapabilityError` before performing a side effect. Idempotent.
    """
    grants._revoked[0] = True
# #EXT-005-REQ-3 End
